tests/ui/c-variadic/valid.rs

```diff
@@ -0,0 +1,59 @@
+//@ run-pass
+#![feature(c_variadic)]
+
+// In rust (and C23 and above) `...` can be the only argument.
+unsafe extern "C" fn only_dot_dot_dot(mut ap: ...) -> i32 {
+    unsafe { ap.arg() }
+}
+
+unsafe extern "C-unwind" fn abi_c_unwind(mut ap: ...) -> i32 {
+    unsafe { ap.arg() }
+}
+
+struct S;
+
+impl S {
+    unsafe extern "C" fn associated_function(mut ap: ...) -> i32 {
+        unsafe { ap.arg() }
+    }
+
+    unsafe extern "C" fn method(&self, mut ap: ...) -> i32 {
+        unsafe { ap.arg() }
+    }
+}
+
+trait T {
+    unsafe extern "C" fn trait_associated_function(mut ap: ...) -> i32 {
+        unsafe { ap.arg() }
+    }
+
+    unsafe extern "C" fn trait_method(&self, mut ap: ...) -> i32 {
+        unsafe { ap.arg() }
+    }
```

Comment:
This raises questions about how we should handle the codegen for its self parameter, dyn compatibility, and so on. I would prefer we not get into those for now and continue to reject this case, please.

Reply:
So reject in traits if there is a `self` parameter of some kind? reject in traits in general?

Previously also just normal associated methods/functions (like in `impl S { }`) were rejected, that is unproblematic right?

Reply:
fair enough, I'll just add a custom error message then.

Reply:
This now generates. At least we have tests now, none of this was exercised apparently.

```
error: associated functions cannot have a C-variadic argument
  --> $DIR/no-associated-function.rs:9:46
   |
LL |     unsafe extern "C" fn associated_function(mut ap: ...) -> i32 {
   |  
```